project/Build.scala

```diff
@@ -1069,6 +1069,49 @@ object Build extends sbt.Build {
   // Testing
 
   val testTagSettings = Seq(
+      javaOptions in Test ++= {
```

Comment:
I guess we don't have the same definition of "common logic" :p

What I'm saying is that both `javaOptions` and `testOptions` perform the same tests to then decide whether or not a "flag" is on or off. For example, regarding `"fastopt-stage"` and `"fullopt-stage"`, in `javaOptions` we have:
```scala
        val stageProp = (scalaJSStage in Test).value match {
          case FastOptStage => Seq(scalaJSProp("fastopt-stage", "true"))
          case FullOptStage => Seq(scalaJSProp("fullopt-stage", "true"))
        }
```
and in `testOptions` we have:
```scala
        val stageTag = Tests.Argument((scalaJSStage in Test).value match {
          case FastOptStage => "-tfastopt-stage"
          case FullOptStage => "-tfullopt-stage"
        })
```
The test and the flags that are produced are the same.

There could be a custom task returning a `Set[String]` (the enabled flags, e.g., `Set("fastopt-stage")`), which is computed based on the common logic. It is then straightforward for `javaOptions` and `testOptions` to convert this `Set[String]` into `-D` options and `Test.Argument`s, respectively.